qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,420 @@
+) -> Tuple[SoquetT, ...]:
+    """Helper function to split / join registers for cirq gates expeciting single 'qubits' register.
+
+    Args:
+        bb: Bloq builder used during decompostion.
+        cirq_bloq: A CirqGateAsBloq wrapped arithmetic gate.
+        regs: bloq registers we wish to use as flat list of qubits for cirq gate.
+
+    Returns:
+        regs: bloq registers appropriately rejoined following split.
+    """
+    flat_regs = []
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            flat_regs.append([v])
+        else:
+            flat_regs.append(bb.split(v))
+    qubits = np.concatenate(flat_regs)
+    qubits = bb.add(cirq_bloq, qubits=qubits)
+    out_soqs = {}
+    start = 0
+    for _, v in regs.items():
+        if v.reg.bitsize == 1:
+            end = start + 1
+            out_soqs[v] = qubits[start:end][0]
+            start += 1
+        else:
+            end = start + v.reg.bitsize
+            out_soqs[v] = bb.join(qubits[start:end])
+            start += v.reg.bitsize
+    return tuple(s for _, s in out_soqs.items())
+
+
+@frozen
+class UniformSuperpositionTHC(Bloq):
+    r"""Prepare uniform superposition state for THC.
+
+    $$
+        |0\rangle^{\otimes 2\log(M+1)} \rightarrow \sum_{\mu\le\nu}^{M} |\mu\rangle|\nu\rangle
+        + \sum_{\mu}^{N/2}|\mu\rangle|\nu=M+1\rangle,
+    $$
+
+    where $M$ is the THC auxiliary dimension, and $N$ is the number of spin orbitals.
+
+    The toffoli complexity of this gate should be 10 * log(M+1) + 2 b_r - 9.
+    Currently it is a good deal larger due to:
+        1. inverting inequality tests should not need more toffolis.
+        2. We are not using phase-gradient gate toffoli cost for Ry rotations
+        3. Small differences in quoted vs implemented comparator costs.
+
+    See: https://github.com/quantumlib/Qualtran/issues/390
+
+    Args:
+        num_mu: THC auxiliary index dimension $M$
+        num_spin_orb: number of spin orbitals $N$
+
+    Registers:
+    - mu: $\mu$ register.
+    - nu: $\nu$ register.
+    - succ: ancilla flagging success of amplitude amplification.
+    - eq_nu_mp1: ancillas for flagging if $\nu = M+1$.
+
+    References:
+        [Even more efficient quantum computations of chemistry through
+            tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf). Eq. 29.
+    """
+
+    num_mu: int
+    num_spin_orb: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("mu", bitsize=(self.num_mu).bit_length()),
+                Register("nu", bitsize=(self.num_mu).bit_length()),
+                Register("succ", bitsize=1),
+                Register("eq_nu_mp1", bitsize=1),
+            ]
+        )
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **regs: SoquetT) -> Dict[str, 'SoquetT']:
+        mu = regs['mu']
+        nu = regs['nu']
+        succ = regs['succ']
+        eq_nu_mp1 = regs['eq_nu_mp1']
```

Comment:
can you just put these in the `build_composite_bloq` signature. My executive decision is that this is nicer despite the mypy warning

Reply:
done.